### PR TITLE
implemented shared and unique ptrs

### PR DESCRIPTION
**Pull request by Srushti24**
Implemented shared and unique pointer.
Use run.sh to run locally

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Use this file to configure clang-format
+
```
Comment (mohitmv) [inline .clang-format:2]: Did you run clang-format on this code ? It doesn't seem to be applied. There are formatting issues in this code.

File: SmartPointers/shared_pointer.hpp [hunk 2]
```diff
@@ -1,80 +1,173 @@
...
+     {
+        T* temp_ptr = m_ptr;
+        T* temp_count = count;
+        m_ptr =  other.m_ptr;
+        count = other.count;
+        other.m_ptr = temp_ptr;
+        other.count = temp_count;
+     }
+    private:
+    T* m_ptr;
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Add a invariant comment here - for both m_ptr and count. And honor that invariant.

Set them nullptr by default directly here , just in case you miss to assign something to them....

File: SmartPointers/shared_pointer.hpp [hunk 3]
```diff
@@ -1,80 +1,173 @@
...
-            (*this->m_count)++;
         }
-        
+        return *this;
     }
-    
-    SharedPointer& operator=(const SharedPointer& sharedPtr) //Copy assignment operator
+
+    // Destructor 
+    ~SharedPointer()
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Create a function named destroy and use that here and other places. Reduce the code repetition.

Also In unique ptr, rename the cleanup to destroy.

File: SmartPointers/shared_pointer.hpp [hunk 4]
```diff
@@ -1,80 +1,173 @@
-//
-//  shared_pointer.hpp
-//  Pointer
-//
-//  Created by Srushti Desai on 12/27/23.
-//
-
-#ifndef shared_pointer_hpp
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp:8]: Why did you remove ifdef ? And comments ?

File: SmartPointers/shared_pointer.hpp [hunk 5]
```diff
@@ -1,80 +1,173 @@
...
+    count(p.count)
+    {
+        (*count)++;
+    }
+
+    //Copy assigment operator
+    SharedPointer& operator=(const SharedPointer& p )
+    {
+         if (this != &p) {
+            *count = *count> 0 ? *count-1: *count;
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Why are you checking it it could be zero already? Is it possible ? Are we maintaining any invariant to prevent that?

File: SmartPointers/shared_pointer.hpp [hunk 6]
```diff
@@ -1,80 +1,173 @@
...
+        return *count;
+     }
     
+    //calculate unique
+     bool unique() const{
+        return *count == 1;
+     }
+
+     //Reset
+     void reset() {
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: In your shared-ptr code, try this:

Sp a = .....

Sp b = std::move(a);

a.reset( );

Your program will crash. It should not crash.

Enable address sanitizer while running your code for capturing memory issues with detailed errors...

File: SmartPointers/shared_pointer.hpp [hunk 7]
```diff
@@ -1,80 +1,173 @@
...
-    }
+        return nullptr;
 
+     }
 
-    
+    //Swap two function
+     void swap(SharedPointer& other)
+     {
+        T* temp_ptr = m_ptr;
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Use std::swap here. All this can be done in 2 lines.

File: src/main.cpp [hunk 8]
```diff
@@ -8,24 +8,38 @@
...
-#include "shared_pointer.hpp"
 #include "unique_pointer.hpp"
+#include "shared_pointer.hpp"
 
 int main(int argc, const char * argv[]) {
-    UniquePointer<int> ptr1;
-    UniquePointer<int>* ptr2 = new UniquePointer<int>();
-    UniquePointer<int> ptr3 = make_unique<int>();
-    std::vector<UniquePointer<int> > testArray;
+    // UniquePointer<int> ptr1;
```
Comment (mohitmv) [inline src/main.cpp]: Add proper unit tests for both shared ptr and unique ptr using assert.... Create two independent helper function for these and call them in main.

File: src/main.cpp
Comment (mohitmv) [inline src/main.cpp:1]: Make a type S with print in ctr and dtr..
Create` vector<UniquePtr<S>>`
Add some item on this vector and checkout the expected behavior.

Comment (mohitmv) [inline src/main.cpp:1]: Create a function that returns UniquePtr..  store the output of the function call on another variable. See what happens

Comment (mohitmv) [inline src/main.cpp:1]: Also try :  pass the output of function directly to another function without storing in a tmp variable.

Comment (mohitmv) [reply src/main.cpp:1]: This is for trying out only. When writing proper unit tests, have some counter in constructor and destructor of S type and assert on their values after using S in unique ptr. Assert on the counter to be done outside the scope where unique ptr is created.

File: SmartPointers/shared_pointer.hpp (hunk 4 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp:8]: Added back

File: SmartPointers/shared_pointer.hpp (hunk 5 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: This is to make sure in case

File: SmartPointers/shared_pointer.hpp (hunk 7 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: resolved

File: SmartPointers/shared_pointer.hpp (hunk 3 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: done

File: SmartPointers/shared_pointer.hpp (hunk 6 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: resolved

File: SmartPointers/shared_pointer.hpp [hunk 9]
```diff
@@ -1,80 +1,134 @@
...
-//  shared_pointer.hpp
-//  Pointer
-//
-//  Created by Srushti Desai on 12/27/23.
-//
+#include <utility>
+#include <iostream>
+#include <stdio.h>
 
-#ifndef shared_pointer_hpp
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp:8]: Why did you remove these `#ifndef shared_pointer_hpp` parts ? They are super critical . Either keep them or use `#pragma once`

File: SmartPointers/shared_pointer.hpp [hunk 10]
```diff
@@ -1,80 +1,134 @@
...
-{
-private:
-    T* m_ptr;
-    unsigned int* m_count;
-public:
-     SharedPointer():m_ptr(nullptr), m_count(new unsigned int(0))
-    {
-        
+  public:
+    // Constructor
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Remove this comment.. Not meaningful.

File: SmartPointers/shared_pointer.hpp [hunk 11]
```diff
@@ -1,80 +1,134 @@
...
-        //cleanup();
-        
+    // Paramatized Constructor
+    SharedPointer(T* ptr) : m_ptr(ptr), count(new int(1)) {
+        std::cout << "called  Shared Paramatized Constructor \n";
     }
 
-    SharedPointer(T * ptr) : m_ptr(ptr), m_count(new unsigned int(1))
-        {
+    // Copy Constructor
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Please remove all of these comments. These comments are extra overhead on brain and don't add any extra information

File: SmartPointers/shared_pointer.hpp [hunk 12]
```diff
@@ -1,80 +1,134 @@
...
+    // Destructor
+    ~SharedPointer() { 
+        std::cout << " Shared Destructor \n";
+        destroy(); }
 
-    SharedPointer(const SharedPointer& sharedPtr){ // Copy constructor
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
+    // Destroy the object
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Remove this comment as well.

You add comments when you need to convey more information than function-name.

File: SmartPointers/unique_pointer.hpp [hunk 13]
```diff
@@ -9,96 +9,77 @@
...
     }
-    return *this;
-}
 };
 
-template<typename T, typename... Args>
-UniquePointer<T> make_unique(Args&&... args)
-{
+template <typename T, typename... Args> UniquePointer<T> make_unique(Args&&... args) {
+    std::cout<< "Unique Pointer make unique called \n";
```
Comment (mohitmv) [inline SmartPointers/unique_pointer.hpp]: Remove std::cout

File: .clang-format (hunk 1 above)
Comment (mohitmv) [reply .clang-format:2]: Done

File: SmartPointers/shared_pointer.hpp (hunk 2 above)
Comment (mohitmv) [reply SmartPointers/shared_pointer.hpp]: You didn't fix this yet.

File: SmartPointers/shared_pointer.hpp (hunk 4 above)
Comment (mohitmv) [reply SmartPointers/shared_pointer.hpp:8]: It was not added in latest commit

File: SmartPointers/shared_pointer.hpp (hunk 6 above)
Comment (mohitmv) [reply SmartPointers/shared_pointer.hpp]: Not resolved properly.

In your shared-ptr code, try this:

Sp a = .....

Sp b = std::move(a);

**a.use_count( );**

Your program will crash. It should not crash.

Enable address sanitizer while running your code for capturing memory issues with detailed errors...

File: SmartPointers/shared_pointer.hpp [hunk 14]
```diff
@@ -1,80 +1,134 @@
...
+            return temp_ptr;
         }
-    
-    int count()
-    {
-        return *m_count;
+        return nullptr;
     }
 
+    // Swap two function
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Why are you swapping two functions ?

File: SmartPointers/shared_pointer.hpp [hunk 15]
```diff
@@ -1,80 +1,134 @@
...
+    void swap(SharedPointer& other) {
+        std::swap(m_ptr, other.m_ptr);
+        std::swap(count, other.count);
+    }
 
-    
+  private:
+    // Raw pointer used to point and count is the counter used to represent number of instances pointing to the same
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp:86]: If you need to add description for these, add 2 comment, one for each of them at their right place ( right above the variable name ).

File: SmartPointers/unique_pointer.hpp [hunk 16]
```diff
@@ -9,96 +9,77 @@
...
 
+    P* ptr() { return this->m_ptr; }
 
-UniquePointer(P* ptr):m_ptr(ptr)
-{
-    
-}
+    void destroy() {
+        delete m_ptr;
+        m_ptr = nullptr; // why do this?
```
Comment (mohitmv) [inline SmartPointers/unique_pointer.hpp]: Remove this line

File: tests/unique_pointer.t.cpp [hunk 17]
```diff
@@ -0,0 +1,23 @@
+#include "unique_pointer.hpp"
+#include <gtest/gtest.h>
+#include <algorithm>
+
+TEST(UniquePointer, creatingNew) {
```
Comment (mohitmv) [inline tests/unique_pointer.t.cpp]: This is good.

Also do it for a custom type S.

File: tests/unique_pointer.t.cpp [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+     UniquePointer<int> uniquePointerOne = make_unique<int>(10);
+     EXPECT_EQ(*uniquePointer, 5);
+     EXPECT_EQ(*uniquePointerOne, 10);
+     uniquePointer = (std::move(uniquePointerOne));
+     EXPECT_EQ(*uniquePointer, 10);
+}
+
+
+TEST(UniquePointer, creatingUniquePointerArray) {
+    std::vector<UniquePointer<int>> uniquePointerArray;
```
Comment (mohitmv) [inline tests/unique_pointer.t.cpp]: Also do it for a custom type S.

Put counters in constructor / destructor of S.

File: tests/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+shared_pointer.t.cpp
+)
+# Create executable for tests
+add_executable(run_tests ${TEST_FILES})
+
+target_include_directories(SmartPointers PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/../SmartPointers)
+
+# Link Google Test to your test executable
+target_link_libraries(run_tests 
+gtest
```
Comment (mohitmv) [inline tests/CMakeLists.txt]: It's ok in this case, but you don't need to necessarily have google-test. You could just use plain `assert` and put them in functions.

**Comment by mohitmv** [comment]
Overall - Remove std::cout completely from ptr implementation. Also remove comments that are not helpful

File: SmartPointers/shared_pointer.hpp [hunk 20]
```diff
@@ -1,80 +1,134 @@
...
-        if(nullptr != sharedPtr.m_ptr)
+    // Destroy the object
+    void destroy() {
+        std::cout << " Shared Destroy Called \n";
+        if(count == nullptr)
         {
-            (*this->m_count)++;
+            return;
+        }
+        *count = *count > 0 ? *count - 1 : *count;
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Let's not have this condition. It should be driven by the invariant

File: SmartPointers/shared_pointer.hpp [hunk 21]
```diff
@@ -1,80 +1,134 @@
...
+
+    // Reset
+    void reset() {
+        destroy();
+        m_ptr = nullptr;
+        count = new int(1);
+    }
+
+    // Paramatized reset
+    void reset(SharedPointer& new_ptr) {
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Your interface seems similar to copy-assignment.

reset method should takes raw pointer in input.

File: SmartPointers/shared_pointer.hpp [hunk 22]
```diff
@@ -1,80 +1,134 @@
...
+        destroy();
+        m_ptr = new_ptr.m_ptr;
+        count = new_ptr.count;
+        *count++;
+    }
+
+    // release
+    T* release() {
```
Comment (mohitmv) [inline SmartPointers/shared_pointer.hpp]: Shared ptr don't have release interface ( Doesn't make sense for shared ptr to release ownership to row ptr, because what will happen to those other objects sharing ownership ? and what is released ptr is manually deleted ? )

https://cplusplus.com/reference/memory/shared_ptr/

File: SmartPointers/shared_pointer.hpp (hunk 5 above)
Comment (mohitmv) [reply SmartPointers/shared_pointer.hpp]: If we had to handle that corner case, there is probably some invariant violation. Let's take good care of invariant in general , so we don't have to handle this rare case possibility.

File: SmartPointers/shared_pointer.hpp (hunk 7 above)
Comment (mohitmv) [reply SmartPointers/shared_pointer.hpp]: This is good. Thanks. ( Just remove the comment from above )

File: SmartPointers/shared_pointer.hpp (hunk 9 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp:8]: solved

File: SmartPointers/shared_pointer.hpp (hunk 10 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: solved

File: SmartPointers/shared_pointer.hpp (hunk 11 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: solved

File: SmartPointers/unique_pointer.hpp (hunk 13 above)
Comment (Srushti24) [reply SmartPointers/unique_pointer.hpp]: solved

File: SmartPointers/unique_pointer.hpp (hunk 16 above)
Comment (Srushti24) [reply SmartPointers/unique_pointer.hpp]: resolved

File: SmartPointers/shared_pointer.hpp (hunk 2 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: fixed

File: SmartPointers/shared_pointer.hpp (hunk 4 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp:8]: fixed

File: SmartPointers/shared_pointer.hpp (hunk 22 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: done

File: SmartPointers/shared_pointer.hpp (hunk 21 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp]: done

File: tests/CMakeLists.txt (hunk 19 above)
Comment (Srushti24) [reply tests/CMakeLists.txt]: done

File: tests/unique_pointer.t.cpp (hunk 18 above)
Comment (Srushti24) [reply tests/unique_pointer.t.cpp]: done

File: tests/unique_pointer.t.cpp (hunk 17 above)
Comment (Srushti24) [reply tests/unique_pointer.t.cpp]: done

File: SmartPointers/shared_pointer.hpp (hunk 15 above)
Comment (Srushti24) [reply SmartPointers/shared_pointer.hpp:86]: done

File: src/main.cpp
Comment (Srushti24) [reply src/main.cpp:1]: done

File: src/main.cpp (hunk 8 above)
Comment (Srushti24) [reply src/main.cpp]: done

File: src/main.cpp
Comment (Srushti24) [reply src/main.cpp:1]: done

Comment (Srushti24) [reply src/main.cpp:1]: done

**Comment by Srushti24** [comment]
I have made all changes except for removing comments, lets keep the comments till i am sure about it.
Its useful for me to understand now what gets called when.

**Comment by Srushti24** [comment]
Invariaent broken :
 p.count = nullptr; - 51 and 61

**Comment by Srushti24** [comment]
I dont think its possible to have an invarient where count is not equal to zero.
when the last shared pointer pointing to an object goes out of scope and its destructor is called, the count should go to zero , as there is nothing left to point to.
I feel there was an error in my previous implementation of move constructor and move assignement operator where i was making count as null, as its move the new owner will point to the object and count will remain same

**Comment by mohitmv** [comment]
1). It seems like you are not running clang-format before pushing a commit. Make it a habit to run clang-format before pushing any commit ( even for temporary stuff ).

2). Why didn't you merge the PR for shared-ptr stuff and create a new PR for shared-pointer-v2 ? ( you called it SharedPtrNull , which is a misleading name )

**Comment by mohitmv** [comment]
Did you get chance to create a google-doc for the things you learnt from previous PR's reviews ?

I see that a lot of mistakes are repeated in PR as well. It would be good to create through each and every review comment in previous PRs and see if there is some generic learning / rules in those comments. If so, document those generic stuff in google-doc, and self review a PR across all the points from that doc, each time you push a new commit.

**Comment by mohitmv** [comment]
Tests:

I see that you wrote literally 2 lines of tests and claimed you implemented shared_ptr_two.
Remember that, these generic and foundational library like shared-ptr like 40% effort in implementing and 60% effort in testing.

For shared-ptr-two, there should be at least **300** lines of tests-code in a independent file, not mixed with unit tests of other utilities. Ideally we should have done it for 1st version of shared_ptr as well but that is fine for now. ( It's ok to copy paste same tests code for 1st version and second-version with minimal differences ).

I like there are a lot of logical mistakes in shared_ptr_two implementation. Most of them needs to be addressed by robust unit testing, covering all possible use cases.

As of now, even the most trivial functionality of shared-ptr are not working. Actually as per the current code, it is not even compiling. After fixing the basic compilation issues myself, it is crashing with illegal-memory-access issues.

```C++
  SharedPointerNull<int> temp;
  {
    auto s1 = SharedPointerNull<int>(new int{4});
    temp = s1;
  }
  std::cout << *temp << "----" << std::endl;
```

**Comment by mohitmv** [comment]
There are some trivial things that we need to fix long before pushing any commit.

1). clang-format.
2). self review code yourself first and find potential issues.
3). checking the code against each learning from previous PRs. ( using the doc you create for learning ).
4). Readable function names and file names. Consistent file naming patterns. Remove meaningless comments from the function - like "// Copy Constructor". Remove comments from the library function before pushing them for code reviews. Feel free to add them for yourself while debugging but remove them before pushing your commit.
